src/snowflake/snowpark/_internal/utils.py

```
@@ -1924,6 +1925,141 @@ def remove_comments(sql_query: str, uuids: List[str]) -> str:
+        raise ValueError(
+            f"Invalid time travel mode: {time_travel_mode}. Must be 'at' or 'before'."
+        )
+
+    # Validate exactly one parameter is provided
+    if time_travel_arg_count != 1:
+        raise ValueError(
+            "Exactly one of 'statement', 'offset', 'timestamp', or 'stream' must be provided."
+        )
+
+    # Validate offset
+    if offset is not None and offset > 0:
+        raise ValueError("'offset' must be a negative integer (seconds).")
+
+    # Normalize timestamp
+    normalized_timestamp = None
+    if timestamp is not None:
+        normalized_timestamp = _normalize_timestamp(timestamp)
+
+    if timezone is not None and timezone.upper() not in ["NTZ", "LTZ", "TZ"]:
+        raise ValueError(
+            f"'timezone' value {timezone} must be None or one of 'NTZ', 'LTZ', or 'TZ'."
+        )
+
+    return TimeTravelConfig(
+        mode=time_travel_mode,
+        statement=statement,
+        offset=offset,
+        timestamp=normalized_timestamp,
+        timezone=timezone,
+        stream=stream,
+    )
+
+
+def _normalize_timestamp(timestamp: Union[str, datetime.datetime]) -> str:
+    """
+    Normalizes timestamp to string format.
+    Args:
+        timestamp: String or datetime object.
+    Returns:
+        Normalized timestamp string.
+    Raises:
+        ValueError: If timestamp format is invalid.
+    """
+    if isinstance(timestamp, datetime.datetime):
+        return str(timestamp)
+    timestamp_str = timestamp.strip()
+    try:
+        # Try to parse with fractional seconds first
+        try:
+            parsed_timestamp = datetime.datetime.strptime(
+                timestamp_str, "%Y-%m-%d %H:%M:%S.%f"
+            )
+        except ValueError:
+            # Fall back to the original format without fractional seconds
+            parsed_timestamp = datetime.datetime.strptime(
+                timestamp_str, "%Y-%m-%d %H:%M:%S"
+            )
+        return str(parsed_timestamp)
```

if users pass us string format timestamp, why we do this str -> datetime -> str conversion for user?

---

using `strptime` here mainly to parse and validate timestamp str format in case of user typo or unsupported format, which could be common since timestamp is strict on format and people have different habits in reporting date & time. would it be better to throw a detailed error here or let the SQL engine handle it in case of str format error?